lib/my_enum.ex

```
@@ -71,7 +71,10 @@ defmodule MyEnum do
     ex) Enum.reverse([1, 2, 3],[4, 5, 6]) => [3, 2, 1, 4, 5, 6]
   """
 
-  def reverse([h | t], tail \\ []), do: reverse(t, [h], tail)
+  def reverse(list, tail \\ [])
+  def reverse([h | t], tail) when is_list(tail), do: reverse(t, [h], tail)
```

타입 체크를 위해 `when` 을 굳이 사용하지 않아도 됩니다.
다른 프로그래밍 언어 같은 경우는 로직상 오류가 발생하면,
프로그램이 터져버려 심각한 피해를 입힐 수 있습니다만,
Erlang OTP / Elixir 에서는 로직이 실패하여도, 가상머신 자체가 터지지않기 때문에
그냥 실패하게 냅두면 됩니다.
iex 콘솔도 하나의 Erlang 프로세스이며, 로직 실패시 프로세스가 터지고, 다시 새로운 프로세스가 생성되어 복구가 되어지게 됩니다.
이걸 "Let It Crash" 라고 하는데 (실패하게 냅둬라),
https://medium.com/pragmatic-programmers/understanding-let-it-crash-f206947f3486 를 참고하면 좋을듯싶네요.

reverse는 아래와 같이 작성할 수 있습니다.
엘릭서 함수를 잘 작성하는 팁은,
마지막 경우의 수를 맨 앞 함수 정의에 다가 두면 쉽게 풀어집니다.
리스트가 빈 [] 인 경우, amount가 0인경우 이런식으로...

```elixir
def reverse([]), do: []
def reverse([h | t]), do: reverse(t) ++ [h]
def reverse(list, tail \\ []), do: reverse(list) ++ tail
```

---

default 정의 부분이 꼭 앞에 올 필요가 없네요 !
 참고해서 다시 수정하겠습니다 !

---

넵 default는 굳이 맨 앞에 둘 필요는 없습니다.
reverse 이외 다른 함수들도 고칠만한게 보이는데 수정해보세용